### PR TITLE
[SMTChecker] Adding support for public getters on `this`

### LINKED ISSUE #10285: [SMTChecker] Support external getters for public state variables

### PR DESCRIPTION
**Pull request by blishko**
This PR adds support for precise analysis of public getters on `this` contract.
Previously these calls were abstracted away in the analysis.
In this PR we treat calls to the getter as an access to the state variable, including the special case of `mapping`s and `array`s where the getter requires parameters corresponding to the index access.

Resolves #10285.

### CODE REVIEW

File: libsolidity/formal/BMC.cpp [hunk 1]
```diff
@@ -478,6 +478,11 @@ void BMC::internalOrExternalFunctionCall(FunctionCall const& _funCall)
 	auto const& funType = dynamic_cast<FunctionType const&>(*_funCall.expression().annotation().type);
 	if (shouldInlineFunctionCall(_funCall))
 		inlineFunctionCall(_funCall);
+	else if (isPublicGetter(&_funCall.expression()))
+	{
+		// DO NOTHING HERE, the processing happens in SMT Encoder, but we need to prevent resetting the state variables
```
Comment (leonardoalt) [inline libsolidity/formal/BMC.cpp]:
```suggestion
		/// Do nothing here.
		/// The processing happens in SMTEncoder, but we need to prevent resetting the state variables.
```

File: libsolidity/formal/BMC.cpp [hunk 2]
```diff
@@ -478,6 +478,11 @@ void BMC::internalOrExternalFunctionCall(FunctionCall const& _funCall)
...
+//		visitPublicGetter(_funCall);
```
Comment (leonardoalt) [inline libsolidity/formal/BMC.cpp]: Can remove this line

File: libsolidity/formal/SMTEncoder.cpp [hunk 3]
```diff
@@ -883,6 +886,102 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+{
+	unsigned dimensionCount = 0;
+	while (_type->category() == Type::Category::Array || _type->category() == Type::Category::Mapping)
+	{
+		++dimensionCount;
+		if (_type->category() == Type::Category::Array)
+			_type = dynamic_cast<ArrayType const *>(_type)->baseType();
+		else if (_type->category() == Type::Category::Mapping)
+			_type = dynamic_cast<MappingType const *>(_type)->valueType();
+		else
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]: We can remove the else, here it's pretty clear it won't happen

File: libsolidity/formal/SMTEncoder.cpp [hunk 4]
```diff
@@ -883,6 +886,102 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+	return dimensionCount;
+}
+
+vector<string> structGetterReturnedMembers(StructType const* _structType)
+{
+	vector<string> returnedMembers;
+	for (auto const& member: _structType->nativeMembers(nullptr))
+	{
+		// TODO: Figure out which types are returned by the getter and which are ignored
+		if (auto category = member.type->category(); category == Type::Category::Array || category == Type::Category::Mapping)
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
		if (
		    auto category = member.type->category();
		    category == Type::Category::Array || category == Type::Category::Mapping
		)
```

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: `string` and `bytes` are fine, right? In that case we'd need an extra condition that if it's an array, it's not `string` nor `bytes`. Also need to check for static arrays

File: libsolidity/formal/SMTEncoder.cpp [hunk 5]
```diff
@@ -2404,6 +2503,13 @@ MemberAccess const* SMTEncoder::isEmptyPush(Expression const& _expr) const
 	return nullptr;
 }
 
+bool SMTEncoder::isPublicGetter(Expression const* _expr) {
+	if (!isTrustedExternalCall(_expr))
+		return false;
+	auto varDecl = dynamic_cast<VariableDeclaration const*>((dynamic_cast<MemberAccess const*>(_expr))->annotation().referencedDeclaration);
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
	auto varDecl = dynamic_cast<VariableDeclaration const*>(
	    dynamic_cast<MemberAccess const*>(_expr)->annotation().referencedDeclaration
	);
```

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: Actually, we first need to check if `_expr` is a `MemberAccess`. I guess now that's guaranteed by `isTrustedExternalCall`, right?

File: libsolidity/formal/SMTEncoder.h [hunk 6]
```diff
@@ -153,6 +153,9 @@ class SMTEncoder: public ASTConstVisitor
 	void visitTypeConversion(FunctionCall const& _funCall);
 	void visitStructConstructorCall(FunctionCall const& _funCall);
 	void visitFunctionIdentifier(Identifier const& _identifier);
+	void visitPublicGetter(FunctionCall const& _funCall);
+
+	bool isPublicGetter(Expression const* _expr);
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.h]:
```suggestion
	bool isPublicGetter(Expression const& _expr);
```

File: libsolidity/formal/SMTEncoder.cpp [hunk 7]
```diff
@@ -883,6 +886,102 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+
+}
+
+void SMTEncoder::visitPublicGetter(FunctionCall const& _funCall)
+{
+	MemberAccess const& access = dynamic_cast<MemberAccess const&>(_funCall.expression());
+	auto var = dynamic_cast<VariableDeclaration const*>(access.annotation().referencedDeclaration);
+	solAssert(var, "");
+	solAssert(m_context.knownExpression(_funCall), "");
+	auto actualArguments = applyMap(_funCall.arguments(), [this](auto const& argExpr) { return expr(*argExpr); });
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
	auto symbArguments = applyMap(_funCall.arguments(), [this](auto argExpr) { return expr(*argExpr); });
```
I think it's good practice to not use & to smart pointer

File: libsolidity/formal/SMTEncoder.cpp [hunk 8]
```diff
@@ -883,6 +886,102 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+	TypePointer type = var->type();
+	auto handleValueType = [&]() {
+		solAssert(actualArguments.empty(), "");
+		m_context.addAssertion(currentValue(*var) == expr(_funCall));
+	};
+	switch (type->category())
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:922]: If you use `type->isValueType()` you could remove all the value type cases, decreasing some lines

File: libsolidity/formal/SMTEncoder.cpp [hunk 9]
```diff
@@ -883,6 +886,102 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+		case Type::Category::Enum:
+		case Type::Category::FixedPoint:
+		case Type::Category::Function:
+		case Type::Category::Contract:
+			handleValueType();
+			break;
+		case Type::Category::Array:
+		case Type::Category::Mapping:
+		{
+			if (auto arrayType = dynamic_cast<ArrayType const*>(type); arrayType && arrayType->isByteArray())
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
			if (
			    auto arrayType = dynamic_cast<ArrayType const*>(type);
			    arrayType && arrayType->isByteArray()
			)
```

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: Shouldn't `string` also go in here?

File: libsolidity/formal/SMTEncoder.cpp [hunk 10]
```diff
@@ -883,6 +886,102 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+				break;
+			}
+			solAssert(dimensionCount(type) == actualArguments.size(), "");
+			smtutil::Expression exprVal = currentValue(*var);
+			for (auto const& arg: actualArguments)
+			{
+				exprVal = smtutil::Expression::select(
+					smtutil::Expression::tuple_get(exprVal, 0),
+					arg
+				);
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:935]: Although this is correct, I think it would be more readable to cast `m_context.variable(*var)` to `SymbolicArrayVariable` and `select` from `symbArray->elements()`

File: libsolidity/formal/SMTEncoder.cpp [hunk 11]
```diff
@@ -883,6 +886,102 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+			}
+			m_context.addAssertion(expr(_funCall) == exprVal);
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
			defineExpr(_funCall, exprVal);
```
Just to make sure the SSA index is increased

File: libsolidity/formal/SMTEncoder.cpp [hunk 12]
```diff
@@ -883,6 +886,102 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+		{
+			auto returnedMembers = structGetterReturnedMembers(dynamic_cast<StructType const*>(type));
+			auto structVar = dynamic_pointer_cast<smt::SymbolicStructVariable>(m_context.variable(*var));
+			auto returnedValues = applyMap(returnedMembers, [&](string const& memberName) { return structVar->member(memberName); });
+			auto const& symbTuple = dynamic_pointer_cast<smt::SymbolicTupleVariable>(m_context.expression(_funCall));
+			solAssert(symbTuple, "");
+			auto const& symbComponents = symbTuple->components();
+			solAssert(symbComponents.size() == returnedValues.size(), "");
+			for (unsigned i = 0; i < symbComponents.size(); ++i)
+			{
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:950]: `{ }` can be removed

File: test/libsolidity/smtCheckerTests/functions/this_external_getter_2.sol [hunk 13]
```diff
@@ -1,12 +0,0 @@
...
-contract C {
-	mapping (uint => uint) public map;
-
-	function f() public view {
-		uint y = this.map(2);
-		assert(y == map[2]); // This fails as false positive because of lack of support for external getters.
-	}
-}
-// ----
-// Warning 6328: (137-156): CHC: Assertion violation happens here.
```
Comment (leonardoalt) [inline test/libsolidity/smtCheckerTests/functions/this_external_getter_2.sol:12]: Why was this file removed?

**Review by leonardoalt — COMMENTED** [review]
Can you also add a test with a state variable of type alternating mapping/array? Not sure what will compile as a getter, but worth a try
```
mapping(uint => uint[]) public m;
mapping(uint => uint[][]) public m;
mapping(uint => uint[][][]) public m;
mapping(uint => mapping(uint => uint[])) public m;
mapping(uint => mapping(uint => uint[][])) public m;
mapping(uint => mapping(uint => mapping(uint => uint[])) public m;

mapping(uint => uint)[] public m;
mapping(uint => uint)[][] public m;
mapping(uint => mapping(uint => uint))[] public m;
mapping(uint => uint[])[] public m;
```

File: libsolidity/formal/BMC.cpp (hunk 2 above)
Comment (blishko) [reply libsolidity/formal/BMC.cpp]: I believe this is related to my hesitance of where to put the call to `visitPublicGetter`.
This seemed like a more natural place for BMC than the one where it called now. But we would have to find a similar good place for CHC.

File: libsolidity/formal/SMTEncoder.cpp (hunk 4 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: True. I added special check for `string` and `byte`. Static arrays are handled the same way as dynamic arrays.

File: libsolidity/formal/SMTEncoder.cpp (hunk 5 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Yes, this is now checked in `isTrustedExternalCall`. Should we duplicate the check here?

File: libsolidity/formal/SMTEncoder.cpp (hunk 9 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Both `bytes` and `string` return `true` for the check `isByteArray`. There is then another method `isString` to distinguish between these further.

File: test/libsolidity/smtCheckerTests/functions/this_external_getter_2.sol (hunk 13 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/functions/this_external_getter_2.sol:12]: It was not removed, just renamed to `this_external_getter_mapping_1.sol ` and updated.
Unfortunately, the file is very small and the git evaluated the change to be more than 50%, so it could not store it as a rename, but it turned it into a deletion of the old file and creation of the new file.

File: libsolidity/formal/SMTEncoder.cpp (hunk 10 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp:935]: That would be nice, the problem is with nested arrays/mappings.
For example for a state variable `uint[][] x` and the call `this.x(1,2)` I have a symbolic variable for `x`, but not for the nested expression x[1].
So what you suggest is only possible for the first iteration of the cycle. For the others, I have only `smtutil::Expression`, but no `SymbolicVariable`.

File: libsolidity/formal/SMTEncoder.cpp (hunk 11 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Changed here and also for the value types. Not sure about the case for the `struct` though.

File: libsolidity/formal/SMTEncoder.cpp (hunk 10 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:935]: Just thought about it again, you're right.

File: libsolidity/formal/SMTEncoder.h (hunk 6 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.h]: Fixed.

File: libsolidity/formal/SMTEncoder.cpp (hunk 7 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Fixed.

File: libsolidity/formal/SMTEncoder.cpp (hunk 8 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp:922]: I've added a similar check.

File: libsolidity/formal/SMTEncoder.cpp (hunk 12 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp:950]: Removed.

File: libsolidity/formal/SMTEncoder.cpp (hunk 3 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Right now that's true. But in case something changes in the future, I would keep it so it's clear. And it will not accidentally end up in an infinite loop if somebody changes the condition but forget to update the body.

File: libsolidity/formal/SMTEncoder.cpp [hunk 14]
```diff
@@ -2403,6 +2505,13 @@ MemberAccess const* SMTEncoder::isEmptyPush(Expression const& _expr) const
 	return nullptr;
 }
 
+bool SMTEncoder::isPublicGetter(Expression const& _expr) {
+	if (!isTrustedExternalCall(&_expr))
+		return false;
+	auto varDecl = dynamic_cast<VariableDeclaration const*>((dynamic_cast<MemberAccess const&>(_expr)).annotation().referencedDeclaration);
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
	auto varDecl = dynamic_cast<VariableDeclaration const*>(
	    dynamic_cast<MemberAccess const&>(_expr).annotation().referencedDeclaration
	);
```

File: libsolidity/formal/SMTEncoder.cpp [hunk 15]
```diff
@@ -867,6 +870,105 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+				auto symbTuple = dynamic_pointer_cast<smt::SymbolicTupleVariable>(m_context.expression(_funCall));
+				solAssert(symbTuple, "");
+				auto const& symbComponents = symbTuple->components();
+				solAssert(symbComponents.size() == returnedValues.size(), "");
+				for (unsigned i = 0; i < symbComponents.size(); ++i)
+					m_context.addAssertion(symbTuple->component(i) == returnedValues.at(i));
+			}
+			break;
+		}
+		default: {} // Unsupported cases,  do nothing and the getter will be abstracted.
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
		default: {} // Unsupported cases, do nothing and the getter will be abstracted.
```

File: libsolidity/formal/SMTEncoder.cpp [hunk 16]
```diff
@@ -867,6 +870,105 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+			_type = dynamic_cast<MappingType const*>(_type)->valueType();
+		else
+			solAssert(false, "");
+	}
+	return dimensionCount;
+}
+
+bool isReturnedFromStruct(TypePointer _type)
+{
+	// So far it seems that only Mappings and ordinary Arrays are not returned (from the cases the compiles).
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]: The code in `ExpressionCompiler::appendStateVariableAccessor` confirms that

File: libsolidity/formal/SMTEncoder.cpp [hunk 17]
```diff
@@ -867,6 +870,105 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
 	m_context.createExpression(_typeName, result.second);
 }
 
+namespace
+{
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:874]: Please add an extra blank

File: libsolidity/formal/SMTEncoder.cpp [hunk 18]
```diff
@@ -867,6 +870,105 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+			_type = dynamic_cast<ArrayType const*>(_type)->baseType();
+		else if (_type->category() == Type::Category::Mapping)
+			_type = dynamic_cast<MappingType const*>(_type)->valueType();
+		else
+			solAssert(false, "");
+	}
+	return dimensionCount;
+}
+
+bool isReturnedFromStruct(TypePointer _type)
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
bool isReturnedFromStructGetter(TypePointer _type)
```

File: libsolidity/formal/SMTEncoder.cpp [hunk 19]
```diff
@@ -867,6 +870,105 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+	auto category = _type->category();
+	if (category == Type::Category::Mapping)
+		return false;
+	if (category == Type::Category::Array)
+		return static_cast<ArrayType const*>(_type)->isByteArray();
+	// default
+	return true;
+}
+
+vector<string> structGetterReturnedMembers(StructType const* _structType)
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
vector<string> structGetterReturnedMembers(StructType const& _structType)
```

File: libsolidity/formal/SMTEncoder.cpp [hunk 20]
```diff
@@ -867,6 +870,105 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+
+}
+
+void SMTEncoder::visitPublicGetter(FunctionCall const& _funCall)
+{
+	MemberAccess const& access = dynamic_cast<MemberAccess const&>(_funCall.expression());
+	auto var = dynamic_cast<VariableDeclaration const*>(access.annotation().referencedDeclaration);
+	solAssert(var, "");
+	solAssert(m_context.knownExpression(_funCall), "");
+	auto symbArguments = applyMap(_funCall.arguments(), [this](auto argExpr) { return expr(*argExpr); });
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
	auto symbArguments = applyMap(_funCall.arguments(), [this](auto argExpr) { return expr(*argExpr, parameterType); });
```
I think we should add the target type for the expression, in case conversion is needed. This might be the case for things like
```
mapping (bytes16 => uint) public m;
...
this.m("cafe");
```
or something like that.
Can you also add a test like that to see what's happening?

For the `parameterType` above, you can call `FunctionType(*var).returnParameterTypes()`

File: libsolidity/formal/SMTEncoder.cpp [hunk 21]
```diff
@@ -867,6 +870,105 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+	TypePointer type = var->type();
+	auto treatAsValueType = [](TypePointer _type) {
+		return _type->isValueType() ||
+			(_type->category() == Type::Category::Array && static_cast<ArrayType const*>(_type)->isByteArray());
+	};
+	if (treatAsValueType(type))
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]: Is the lambda used anywhere else? If not I think it's better to just inline the condition here

File: libsolidity/formal/SMTEncoder.cpp [hunk 22]
```diff
@@ -867,6 +870,105 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+					arg
+				);
+			}
+			defineExpr(_funCall, exprVal);
+			break;
+		}
+		case Type::Category::Struct:
+		{
+			auto returnedMembers = structGetterReturnedMembers(dynamic_cast<StructType const*>(type));
+			solAssert(!returnedMembers.empty(), "");
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:943]: Interesting, didn't know this was the case.

File: libsolidity/formal/SMTEncoder.cpp (hunk 16 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: I see it as well. I wonder if it can be refactored to a method that can be called both from `ExpressionCompiler` and from `SMTEncoder`.

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: I think it's not worth extracting it, let's leave it as it is

File: libsolidity/formal/SMTEncoder.cpp (hunk 20 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Good catch! Without the type conversion, this crashes now for the test you suggested. I will fix it and add the test case.

File: libsolidity/formal/SMTEncoder.cpp (hunk 21 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Inlined.

File: libsolidity/formal/SMTEncoder.cpp (hunk 19 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Replaced.

File: libsolidity/formal/SMTEncoder.cpp (hunk 18 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Renamed.

File: libsolidity/formal/SMTEncoder.cpp (hunk 17 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp:874]: Added.

File: libsolidity/formal/SMTEncoder.cpp (hunk 15 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Fixed.

File: libsolidity/formal/SMTEncoder.cpp (hunk 14 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Fixed.

File: libsolidity/formal/SMTEncoder.cpp [hunk 23]
```diff
@@ -867,6 +870,93 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
 	m_context.createExpression(_typeName, result.second);
 }
 
+namespace // helpers for SMTEncoder::visitPublicGetter
+{
+
+bool isReturnedFromStructGetter(TypePointer _type)
+{
+	// So far it seems that only Mappings and ordinary Arrays are not returned (from the cases the compiles).
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
	// So far it seems that only Mappings and ordinary Arrays are not returned.
```

File: libsolidity/formal/SMTEncoder.cpp [hunk 24]
```diff
@@ -867,6 +870,93 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+	auto category = _type->category();
+	if (category == Type::Category::Mapping)
+		return false;
+	if (category == Type::Category::Array)
+		return static_cast<ArrayType const*>(_type)->isByteArray();
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
		return dynamic_cast<ArrayType const&>(_type).isByteArray();
```
`static_cast` _should_ be fine, but we favor `dynamic_cast` because of paranoia about crashes, so I think we should stick to it.

File: libsolidity/formal/SMTEncoder.cpp [hunk 25]
```diff
@@ -867,6 +870,93 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+	return true;
+}
+
+vector<string> structGetterReturnedMembers(StructType const& _structType)
+{
+	vector<string> returnedMembers;
+	for (auto const& member: _structType.nativeMembers(nullptr))
+		if (isReturnedFromStructGetter(member.type))
+			returnedMembers.push_back(member.name);
+	return returnedMembers;
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:894]: This is a good candidate for a `filter` function.

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:894]: (in the future)

File: libsolidity/formal/SMTEncoder.cpp [hunk 26]
```diff
@@ -867,6 +870,93 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+	auto actualArguments = _funCall.arguments();
+	solAssert(actualArguments.size() == paramExpectedTypes.size(), "");
+	vector<smtutil::Expression> symbArguments;
+	for (unsigned i = 0; i < paramExpectedTypes.size(); ++i)
+		symbArguments.push_back(expr(*actualArguments[i], paramExpectedTypes[i]));
+
+	TypePointer type = var->type();
+	if (
+		type->isValueType() ||
+		(type->category() == Type::Category::Array && static_cast<ArrayType const*>(type)->isByteArray())
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
		(type->category() == Type::Category::Array && dynamic_cast<ArrayType const&>(type).isByteArray())
```

File: libsolidity/formal/SMTEncoder.cpp [hunk 27]
```diff
@@ -867,6 +870,93 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+	switch (type->category())
+	{
+		case Type::Category::Array:
+		case Type::Category::Mapping:
+		{
+			smtutil::Expression exprVal = currentValue(*var);
+			for (auto const& arg: symbArguments)
+			{
+				exprVal = smtutil::Expression::select(
+					smtutil::Expression::tuple_get(exprVal, 0),
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:933]: Maybe add a comment here about what it's doing? It's clear to us, but probably less clear to other readers

File: libsolidity/formal/SMTEncoder.cpp [hunk 28]
```diff
@@ -867,6 +870,93 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+					arg
+				);
+			}
+			defineExpr(_funCall, exprVal);
+			break;
+		}
+		case Type::Category::Struct:
+		{
+			auto returnedMembers = structGetterReturnedMembers(static_cast<StructType const&>(*type));
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
			auto returnedMembers = structGetterReturnedMembers(dynamic_cast<StructType const&>(*type));
```

File: libsolidity/formal/SMTEncoder.cpp [hunk 29]
```diff
@@ -867,6 +870,93 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+			solAssert(!returnedMembers.empty(), "");
+			auto structVar = dynamic_pointer_cast<smt::SymbolicStructVariable>(m_context.variable(*var));
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:944]: I think here we should either `solAssert(structVar, "")` or change the cast to `SymbolicStructVariable const&`

File: libsolidity/formal/SMTEncoder.cpp [hunk 30]
```diff
@@ -867,6 +870,93 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+			auto returnedValues = applyMap(returnedMembers, [&](string const& memberName) { return structVar->member(memberName); });
+			if (returnedValues.size() == 1)
+				defineExpr(_funCall, returnedValues.front());
+			else
+			{
+				auto symbTuple = dynamic_pointer_cast<smt::SymbolicTupleVariable>(m_context.expression(_funCall));
+				solAssert(symbTuple, "");
+				auto const& symbComponents = symbTuple->components();
+				solAssert(symbComponents.size() == returnedValues.size(), "");
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:955]: I think better increase the struct SSA index to be safe

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:955]: Ah, it would need to be updated before `returnedValues` is defined

File: libsolidity/formal/SMTEncoder.cpp [hunk 31]
```diff
@@ -867,6 +870,93 @@ void SMTEncoder::endVisit(ElementaryTypeNameExpression const& _typeName)
...
+}
+
+vector<string> structGetterReturnedMembers(StructType const& _structType)
+{
+	vector<string> returnedMembers;
+	for (auto const& member: _structType.nativeMembers(nullptr))
+		if (isReturnedFromStructGetter(member.type))
+			returnedMembers.push_back(member.name);
+	return returnedMembers;
+}
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:895]: Please add a blank line here

**Review by leonardoalt — COMMENTED** [review]
@blishko a few more comments, but then it should be good

File: libsolidity/formal/SMTEncoder.cpp (hunk 25 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp:894]: Definitely. Are there any plans for that?

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:894]: C++20 I guess :p
Let's just leave it for now

File: libsolidity/formal/SMTEncoder.cpp (hunk 30 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp:955]: I don't think we can do that.
I added a test this morning exactly for situation like that and it seems this would break it.

Basically this getter is read-only operation. So if I call it twice in a row, it should be provable that I read the same thing. While if we increase the index *before* creating the `returnedValues` it would break this.
I am not sure now if we can increase the index *after* that.

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:955]: Ah right, it's the other way around, I think we should increase the index of `symbTuple`, that's also what the `if` branch does via `defineExpr`

Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp:955]: We can increase the index on the struct SSA at the end of the struct `case`. Do you think we should do that?

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:955]: No I think you're right that the struct is read only and we shouldn't change its index, what I mean is getting a fresh index to `_funCall` before assigning to it

Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp:955]: Done.

File: libsolidity/formal/SMTEncoder.cpp (hunk 29 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp:944]: `solAssert` added.

File: libsolidity/formal/SMTEncoder.cpp (hunk 27 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp:933]: I added a comment, not sure I can phrase it well what is going on there...

File: libsolidity/formal/SMTEncoder.cpp (hunk 31 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp:895]: Added.

**Review by leonardoalt — COMMENTED** [review]
Looks good. Please squash & rebase

**Comment by blishko** [comment]
Squashed and rebased.

**Review by leonardoalt — APPROVED** [review]
